lib/tds.js

```diff
@@ -290,19 +310,29 @@ async function generateDNRRulesForTrackerEntry (
         }
 
         priority += TRACKER_RULE_PRIORITY_INCREMENT
-        dnrRules.push(
-            generateDNRRule({
-                priority,
-                actionType: ruleAction,
-                redirect: redirectAction,
-                urlFilter,
-                regexFilter,
-                matchCase,
-                requestDomains,
-                excludedInitiatorDomains,
-                resourceTypes
-            })
-        )
+
+        const newRuleTemplate = {
+            priority,
+            actionType: ruleAction,
+            redirect: redirectAction,
+            urlFilter,
+            regexFilter,
+            matchCase,
+            requestDomains,
+            excludedInitiatorDomains,
+            resourceTypes
+        }
+        const newRule = generateDNRRule(newRuleTemplate)
+
+        if (customRuleAction) {
+            // if a custom action, map the action to the specified default
+            // and add its inverse to the list of custom action rules
+            const inverseRule = { ...newRuleTemplate }
+            inverseRule.actionType = customRuleAction.inverseAction
+            newRule[inverseCustomActionSymbol] = { customAction: trackerEntryRules[i].action, ...generateDNRRule(inverseRule) }
```

Comment:
How about we assign the custom action here, then let the finalize step do the rest?
```suggestion
            newRule[customActionSymbol] = customRuleAction
```

Reply:
I actually disagree.  This method is `generateDNRRulesForTrackerEntry`, and the new rule is generated right above via `newRuleTemplate`, and so its inverse should be generated here as well.

Otherwise I'd have to copy or recreate that template in the finalize step, which means we're muddling where rule generation happens and run the risk of drift over time where the inverse may no longer have the same properties as the original.  Or I go back to cloning the generated DNR rule in finalize.  

Already rewrote this once, I think we should leave it be unless there's a specific issue to fix?

Reply:
Well, I think it would make more sense to let finalize step take care of it. To my mind, this is kind of meta-data for the rule, that it has this custom action that also needs to be handled. The rule itself and the custom action is all that's needed to generate the inverse rule (or rules for different kinds of custom actions in the future). It's not known if the inverse rule (or whatever results from other sorts of custom actions) is required until the finalize step.

I don' t think implementing it the way I'm suggesting would be too hard. Here we'd do something like:
```javascript
let customRuleAction = null
if (supportedCustomActions.has(ruleAction)) {
    customRuleAction = ruleAction
    ruleAction = supportedCustomActions.get(ruleAction).action
}

...

if (customRuleAction) {
    newRule[customActionSymbol] = customRuleAction
}
```

and then in the finalize function, we'd do something like:

```javascript
const inverseRulesByCustomAction = new Map()

...

const customAction = rule[customRuleActionSymbol]
delete  rule[customRuleActionSymbol]
if (customAction) {
    const customActionDetails = supportedCustomActions.get(customAction)
    
    if (customActionDetails.inverseAction) {
        const inverseRule = JSON.parse(JSON.stringify(rule))
        storeInLookup(inverseRulesByCustomAction, customAction, [inverseRule])
    }
}

...

return { ruleset, matchDetailsByRuleId, inverseRulesByCustomAction }

```

Reply:
I think we need to do a call about this, I'm available until about 9am your time or again your evening.